tests/rptest/tests/transactions_test.py

```diff
@@ -244,3 +244,70 @@ def change_static_member_test(self):
             assert kafka_error.code() == ck.cimpl.KafkaError.FENCED_INSTANCE_ID
 
         producer.abort_transaction()
+
+
+class UpgradeTransactionTest(RedpandaTest):
+    @cluster(num_nodes=2, log_allow_list=RESTART_LOG_ALLOW_LIST)
+    def check_parsing_test(self):
+        self.redpanda = RedpandaService(
+            self.test_context,
+            1,
+            extra_rp_conf={
+                "enable_idempotence": True,
+                "enable_transactions": True,
+                "transaction_coordinator_replication": 1,
+                "id_allocator_replication": 1,
+                "enable_leader_balancer": False,
+                "enable_auto_rebalance_on_node_add": False,
+            },
+            environment={"__REDPANDA_LOGICAL_VERSION": 3})
+
+        self.redpanda.start()
+
+        spec = TopicSpec(partition_count=1, replication_factor=1)
+        self._client = DefaultClient(self.redpanda)
+        self.client().create_topic(spec)
+        topic_name = spec.name
+
+        producer = ck.Producer({
+            'bootstrap.servers': self.redpanda.brokers(),
+            'transactional.id': '123',
+            'transaction.timeout.ms': 10000,
+        })
+
+        producer.init_transactions()
+
+        def on_del(err, msg):
+            assert err == None
+
+        max_tx = 100
+        for i in range(max_tx):
+            producer.begin_transaction()
+            producer.produce(topic_name, str(i), str(i), 0, on_del)
+            producer.commit_transaction()
+
+        self.redpanda.set_environment({"__REDPANDA_LOGICAL_VERSION": 4})
+        for n in self.redpanda.nodes:
+            self.redpanda.restart_nodes(n, stop_timeout=60)
+
+        consumer = ck.Consumer({
```

Comment:
This doesn't access the code path of tm_stm serialization does it? (curious how its testing any code paths in this patch because after L287 we have nothing to do with tm_stm) ..

Reply:
It access tm_stm serialization, but for old format. I have a plan to add  normal upgrade test